Feature/preset configurator

seb5g opened the pull request:
Fix #754
 
Ideas of ameliorations for the configurator

See #748 for details on the ideas

- [ ]  list view or a QCombobox
- [x] delegates styling
- [x] automatic saving
- [x] move up/down buttons
- [x] waiting time entry
- [x] context menu styling
- [ ] buttons for specific actions a bit like in the preset to select control modules
- [x] A visual indicator to know what can be "configured" would be great
- [x] readonly settings
- [x] The ... was experimental to be quick in creating a new one. The ... is normally never saved.
- [x] shortcuts?
- [x] doucleclick

seb5g committed:
dialog not showing (sometimes) if None widget is added to layout??

seb5g committed:
patch to accept older style preset_files (without applying complex settings)

seb5g committed:
accepting an applied setting if its path is valid!

seb5g committed:
removing the misleading '...' entry (redundant with +)

seb5g committed:
all input settings are set as readonly (cannot change their value)

but those who were readonly cannot be configured: their option VALID_FOR_CONFIGURATION is set to False

seb5g committed:
if one applies an entry: it is automatically saved

if one

seb5g committed:
handling Special Entries using a Factory

seb5g committed:
done with the three special entries

seb5g committed:
color in green settings that are valid to be selected for configuration

seb5g committed:
allow multiple selection and move and delete of several rows

seb5g committed:
move down and up buttons trigger correct action

seb5g committed:
double click and add button (right arrow) split grouped parameterrs

seb5g commented:
Hi @Ashwolaa I implemented most of your suggestions. Do you think you could suggest something for the delegate widgets? The button for the specific actions (but check the context menu first)? and what shortcuts would be useful?

seb5g commented:
@malik-irain the tests are failing because the pymodaq_utils package I modified is not installed for the pymodaq tests (it installs pymodaq_utils from pypi and not from the package folder...). Why not using your installer?

Ashwolaa commented:
I am doing this test after firing the install-packages.py. I have not really looked at the dev branch yet.

Here are a few comments on the usage.

From the menu in the dashboard:
- Pressing Preset Manager / Action Manager should show up the window in front of the current ones (like for settings). 
- When reloading a preset, the old window is still there. It can be closed though.
<img width="619" height="170" alt="image" src="https://github.com/user-attachments/assets/09935846-1d08-4c18-b8b6-fd656a926cbb" />


In the configurator:
- When added with double click, the drag and drop does not work inside the table
- I receive some errors when I try to apply the configurator. For each item, I have a Value error saying that it is not a supported entry. I did not manage to get it to work.
```
Traceback (most recent call last):
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\utils.py", line 216, in <lambda>
    self.connect_action(ManagerActions.APPLY, lambda: self.apply_entry_base())
                                                      ~~~~~~~~~~~~~~~~~~~~~^^
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\utils.py", line 320, in apply_entry_base
    self.apply_entry(entry_path, **kwargs)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\configurator\configurator.py", line 115, in apply_entry
    incompatible_index = self.check_parameters(pwp_list, settings_all)
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\configurator\configurator.py", line 149, in check_parameters        
    special_entry = special_entry_factory.get_entry_from_long_name(entry.setting.path[-1])(
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\configurator\special_entries.py", line 147, in get_entry_from_long_e
    raise ValueError(f".{special_entry_name_long} is not a supported entry.")
ValueError: .scaling is not a supported entry.
```
-  the plus button should show up a fresh parametertree if a name is given
- when pressing the minus button, the combobox shows the last preset in the list but the parameter tree is not refreshed with the correct values.
- items can only be added once currently from the parameter tree. Is that intended? 
- items added from special configuration cannot be dragged/dropped
- the DEL key erases all the selected entries. This should be the case with the right-click (remove selected rows action)
- init false is not working in the special actions 
- wait time special action leads to:
```
Traceback (most recent call last):
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\utils.py", line 216, in <lambda>
    self.connect_action(ManagerActions.APPLY, lambda: self.apply_entry_base())
                                                      ~~~~~~~~~~~~~~~~~~~~~^^
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\utils.py", line 320, in apply_entry_base
    self.apply_entry(entry_path, **kwargs)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\configurator\configurator.py", line 124, in apply_entry
    special_entry.apply_entry(entry, module=mod, dashboard=self.dashboard)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\constant.schouder\Documents\PyMoDAQ\PyMoDAQ\packages\pymodaq\src\pymodaq\utils\managers\configurator\special_entries.py", line 294, in apply_entry
    QtCore.QThread.msleep(entry.setting.value() * 1000)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: msleep(a0: int): argument 1 has unexpected type 'float'
```
- I am not so fond of the pop up windows which just says "Applying Configuration: ".  I would rather follow in real time on the configurator entry if the action is successful or not. This is what I had in mind when adding an extra column that turns red/green according to the status. Or at least have some information on which things it is currently doing. This pop up window stays open when it crashes by the way. A catch mechanism is missing.
- This green color is "okayish". How about a different approach, we only show the settings that can be changed! We can just add a checkbox on top (in the parameter tree itself?) which would display the full settings with the color green/red for those who can be used in the configurator entry.
- the main in configurator.py does not work currently (it loads the preset but I cannot drqg/drop items from it)

Some shortcuts:
- ctrl + s ==) Save ; ctrl +r ==) Reload; ctrl +e ==) Execute; ctrl + q ==) Quit

seb5g committed:
better configuration of Parameter delegate to fill the size of the cell

seb5g committed:
applying hide/show valid settings

seb5g committed:
raised error if already existing (normally checked before but could be raised if working with parallel node (pytest)

seb5g committed:
change display settings given their configurability!

seb5g committed:
change installed packages for tests

seb5g committed:
Update tests-pymodaq.yml

seb5g committed:
remove direct pyqt6 import

seb5g committed:
Update tests-pymodaq.yml

seb5g committed:
Update tests-pymodaq.yml

seb5g committed:
Update tests-pymodaq.yml

seb5g committed:
make sure the Manager is on top when clicking its creation/show button

seb5g committed:
Update tests-pymodaq.yml

seb5g commented:
> @malik-irain the tests are failing because the pymodaq_utils package I modified is not installed for the pymodaq tests (it installs pymodaq_utils from pypi and not from the package folder...). Why not using your installer?

@malik-irain :
I worked on that also. It works as intended but all the tests for python 3.9 are failing, related to pytables (which is there as I checked the installed dependencies)...

<img width="1447" height="773" alt="image" src="https://github.com/user-attachments/assets/67031c40-5931-497c-99f3-a3ea1452396d" />

seb5g commented:
@malik-irain and there is some random errors on PArameter tests in pymodaq_gui?? Surprisingly, the only tests that are passing are the ones for python 3.9...

seb5g commented:
> I am doing this test after firing the install-packages.py. I have not really looked at the dev branch yet.
> 
> Here are a few comments on the usage.
> 
> From the menu in the dashboard:
> 
>     * Pressing Preset Manager / Action Manager should show up the window in front of the current ones (like for settings).

Done: The manager is raised!

> 
>     * When reloading a preset, the old window is still there. It can be closed though.

Managed to get rid of it by calling processEvents after each module/dock/widget closing
> 
> 

> In the configurator:
> 
>     * When added with double click, the drag and drop does not work inside the table

This is due to failure in the settings compare methods, we have to discuss this... !!!
It now works!

> 
>     * I receive some errors when I try to apply the configurator. For each item, I have a Value error saying that it is not a supported entry. I did not manage to get it to work.
> 
> 
> ```

> ValueError: .scaling is not a supported entry.
> ```
done the Special entry had to raise a KeyError to trigger proper behaviour...

> 
>     * the plus button should show up a fresh parametertree if a name is given
done

> 
>     * when pressing the minus button, the combobox shows the last preset in the list but the parameter tree is not refreshed with the correct values.

it is now refreshed
> 
>     * items can only be added once currently from the parameter tree. Is that intended?
well yes this is inforced as there is no reason for a given entry to have several same configentry. Be carefull not to mix things with the future sequencer that will allow to aply several config one after the other (for that purpose, the waiting time is a also a bit out of order here...) Especially as now the move_actuator is blocking (so make sure to make several moves in the right order)
> 
>     * items added from special configuration cannot be dragged/dropped

it is now (same reason as for the double clicked one I guess)
> 
>     * the DEL key erases all the selected entries. This should be the case with the right-click (remove selected rows action)
> 

Done

>     * init false is not working in the special actions
it is now

> 
>     * wait time special action leads to:
> 

>     ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> TypeError: msleep(a0: int): argument 1 has unexpected type 'float'
> ```
> 
corrected and also call processevent

>     * I am not so fond of the pop up windows which just says "Applying Configuration: ".  I would rather follow in real time on the configurator entry if the action is successful or not. This is what I had in mind when adding an extra column that turns red/green according to the status. Or at least have some information on which things it is currently doing. This pop up window stays open when it crashes by the way. A catch mechanism is missing.

ongoing
> 
>     * This green color is "okayish". How about a different approach, we only show the settings that can be changed! We can just add a checkbox on top (in the parameter tree itself?) which would display the full settings with the color green/red for those who can be used in the configurator entry.

done
> 
>     * the main in configurator.py does not work currently (it loads the preset but I cannot drqg/drop items from it)

now it loads a saved settings.xml from the tests folder (for testing only obviously as it has no meaning to have the configurator stand alone)
> 
> 
> Some shortcuts:
> 
>     * ctrl + s ==) Save ; ctrl +r ==) Reload; ctrl +e ==) Execute; ctrl + q ==) Quit

ongoing

seb5g committed:
get rid of some docks that were lingering when reapplying another preset

seb5g committed:
double clicking an entry => i can drag it after (was not working with some issue in parameter comparison)

seb5g committed:
when deleting an entry, update the representation corresponding to the new entry name showed in the combobox

seb5g committed:
context menu remove selected remove all selevted rows

seb5g committed:
set the raised error of the factory to a KeyError and cast waiting time to int in order to use msleep

seb5g committed:
correct init or desinit modules. Proper waiting time with kind of non blocking behviour

seb5g committed:
standart config entry for settings also in factory

seb5g committed:
all entries type are working. Ready to create steps!

seb5g committed:
created a widget and listview to display the entries when they are applied and when they fail!

seb5g committed:
updated documents

seb5g committed:
implemented base class to display initialization of manager sublist application

seb5g committed:
apply the entry_sublist display to preset manager

seb5g committed:
replaced splashscreen with custom widget displaying entry sublist loading

seb5g committed:
adding title to list and adapting Font

seb5g committed:
Merge branch 'bugfix/actuators_detectors_in_module_manager' into feature/preset_configurator

seb5g committed:
add shortcuts

seb5g committed:
bugfix: setter actuators_all didn't synchronize with settings

seb5g commented:
Everything (and more) is implemented!! 


https://github.com/user-attachments/assets/5bb91f20-2635-45a4-addb-a8c4bb0dca9f

seb5g commented:
@Ashwolaa ready for another round of review! And then I'll end with the documentation

Ashwolaa commented:
- When the action fails (moving an actuator when it is not init), the status stays yellow but goes green after the Error Timeout
- Accessing the value of a boolean item in the table leads to:
  - ` QAbstractItemView::commitData called with an editor that does not belong to this view`
- Moving booleans lead to the `TypeError: invalid result from ConfiguratorModel.moveRows(), a 'bool' is expected not 'NoneType'`
- Managed to make it crash! Tough one, but after you add multiple similar entries with the insert special value
  - The program does not like it when you try to move them around. 
  - I could make it happen by adding two init (true then false) and one motion between. Trying to move an init makes the dashboard crashed.
  - This only happens with drag/drop (Ctrl+up/down does not affect it)
- double click does not work? Neither right arrow (from the GUI)
Shortcuts:
- I prefered ctrl + E to be consistent with the DAQ_Move but you're the boss
- ctrl + N for a new config/preset?

I try to find some extra time to finish it

seb5g committed:
remove poster files

seb5g committed:
renamed objects/attributes to reflect the subentry type and the subentry handler type

handle also the SubEntryError raising in case an issue happens within the handler apply_subentry

seb5g committed:
Merge branch 'dev' into feature/preset_configurator

seb5g committed:
renamed module special_entries to subentries

seb5g committed:
handle subentry that has no control module associated with

seb5g committed:
moving things within the same table is directly handled by a call to moveRows! Make sure there are no duplicates in subentries

seb5g committed:
adding setting (with arrow) and double clicking is working again

seb5g committed:
changed the apply word to execute with ctrl+E shortcut (just like for executing absolute moves in daq_move)

seb5g committed:
fixing long string

seb5g commented:
> * When the action fails (moving an actuator when it is not init), the status stays yellow but goes green after the Error Timeout
> 
I added handling of potential errors and correctly displaying the text as red

>     * Accessing the value of a boolean item in the table leads to:
>       * ` QAbstractItemView::commitData called with an editor that does not belong to this view`
could not reproduce this error!

>     * Moving booleans lead to the `TypeError: invalid result from ConfiguratorModel.moveRows(), a 'bool' is expected not 'NoneType'`

reimplemented the moveRows, it solved on the way the issue just below

> 
>     * Managed to make it crash! Tough one, but after you add multiple similar entries with the insert special value
>       
>       * The program does not like it when you try to move them around.
>       * I could make it happen by adding two init (true then false) and one motion between. Trying to move an init makes the dashboard crashed.
>       * This only happens with drag/drop (Ctrl+up/down does not affect it)

it's ok now, but anyway I also make sure one cannot add two same subentry. A configuration entry is a list of subentries being a particular state of the dashboard. If one want to apply different configuration, one has to wait for the Sequencer ;-)

> 
>     * double click does not work? Neither right arrow (from the GUI)
back in working order

>       Shortcuts:
>     * I prefered ctrl + E to be consistent with the DAQ_Move but you're the boss
that's true that I'm the boss, but I'm listening to the subgourou and applied (executed) that in the code :-D
>     * ctrl + N for a new config/preset?
Done
> 
> 
> I try to find some extra time to finish it

In the mean time I move on to documentation

seb5g committed:
using full enum as entry_type

seb5g committed:
Update configurator_manager_test.py

seb5g committed:
updated doc on the preset manager and configurator

seb5g committed:
update doc

seb5g commented:
all tests pass. I'll merge this PR and we'll patch it if necessary while using it!
